tinysl_test.go

```diff
@@ -4,277 +4,416 @@ import (
-	t.Run("Service can be registered as Transient", testAdd(Transient))
-	t.Run("Service can be registered as PerContext", testAdd(PerContext))
-	t.Run("Service can be registered as Singleton", testAdd(Singleton))
-	t.Run("Service cannot be registered twice", testAddTwice)
-	t.Run("Service cannot be registered twice regardless of life time", testAddTwiceDifferentLifetime)
-	t.Run("Same implementation can be registered for 2 different interfaces",
-		testSameImplementationDifferentServices)
-	t.Run("Transient Service is always returned as a new instance", testTransientNewInstance)
-	t.Run("PerContext Service instance is same per Context", testPerContextSameContext)
-	t.Run("PerContext Service instance is different for different Contexts",
-		testPerContextDifferentContext)
-	t.Run("PerContext Service instance is not returned for cancelled Context",
-		testPerContextCancelledContext)
-	t.Run("Singleton Service instance is always the same regardless of Context",
-		testSingletonSameInstance)
-	t.Run("Can pass constructor with Context argument", testPassConstructorWithContext)
-	t.Run("Cannot pass constructor with more than 1 argument or argument not of type Context",
-		testCanPassNoArgsOrWithContextConstructor)
-	t.Run("Can pass constructor with Context argument only with PerContext lifetime",
-		testPassConstructorWithContextOnlyToPerContext)
-	t.Run("Can use ServiceLocator inside Transient services",
-		testCanUseSLInsideTransientServices)
-	t.Run("Can use ServiceLocator inside PerContext services",
-		testCanUseSLInsidePerContextServices)
-	t.Run("Can use ServiceLocator inside Singleton services",
-		testCanUseSLInsideSingletonServices)
+	suite.Run(t, new(serviceRegistrationSuite))
+	suite.Run(t, new(serviceCreationSuite))
 }
 
-func testAdd(lifetime lifetime) func(*testing.T) {
-	return func(t *testing.T) {
-		assert := assert.New(t)
+type serviceRegistrationSuite struct {
+	suite.Suite
+}
 
-		ls := New()
-		i := new(int)
-		err := ls.Add(lifetime, getServiceC(i))
+func (suite *serviceRegistrationSuite) testAdd(lifetime lifetime) {
+	ls := New()
+	i := new(int)
+	err := ls.Add(lifetime, getServiceC(i))
 
-		assert.NoError(err, "should not return any error")
-	}
+	suite.NoError(err, "should not return any error")
 }
 
-func testAddTwice(t *testing.T) {
-	assert := assert.New(t)
+func (suite *serviceRegistrationSuite) TestAddSingleton() {
+	suite.testAdd(Singleton)
+}
 
+func (suite *serviceRegistrationSuite) TestAddPerContext() {
+	suite.testAdd(PerContext)
+}
+
+func (suite *serviceRegistrationSuite) TestAddTransient() {
+	suite.testAdd(Transient)
+}
+
+func (suite *serviceRegistrationSuite) TestAddDuplicate() {
 	ls := New()
 	i := new(int)
 
 	err := ls.Add(Transient, getServiceC(i))
-	assert.NoError(err, "should not return any error")
+	suite.NoError(err, "should not return any error")
 
 	err = ls.Add(Transient, getServiceC(i))
-	assert.Error(err, "should return an error")
+	suite.Error(err, "should return an error")
```

Comment:
In general, I would advise to always check the error you get back with the one you expect to avoid a test passing with an unexpected error being returned.

Reply:
I agree, but it would be better to do it in separate pull-request I think